sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/scopes.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.catalyst.analysis.{Resolver, UnresolvedAttribute}
+import org.apache.spark.sql.catalyst.plans.logical.{CoGroup, LogicalPlan, MapGroups}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types.DataType
+
+
+/**
+ * An expression that has to be resolved against a scope of resolved attributes.
+ */
+case class ScopedExpression(expr: Expression, scope: Seq[Attribute])
+  extends Expression with Unevaluable {
+  override def children: Seq[Expression] = expr +: scope
+  override def dataType: DataType = expr.dataType
+  override def nullable: Boolean = expr.nullable
+  override def prettyName: String = "scoped"
+  override def sql: String = s"$prettyName(${expr.sql}, $scope)"
+  override lazy val resolved: Boolean = expr.resolved
+
+  override protected def withNewChildrenInternal(children: IndexedSeq[Expression]): Expression = {
+    val scope = children.tail
+    assert(scope.forall(_.isInstanceOf[Attribute]), "Scope children have to be attributes")
+    copy(expr = children.head, scope = scope.map(_.asInstanceOf[Attribute]))
+  }
+}
+
+/**
+ * Restricts the scope of resolving some expressions.
+ */
+object ScopeExpressions extends Rule[LogicalPlan] {
+  private def scopeOrder(scope: Seq[Attribute])(sortOrder: SortOrder): SortOrder = {
+    sortOrder match {
+      case so if so.child.isInstanceOf[ScopedExpression] => so
+      case so => so.copy(
+        child = ScopedExpression(so.child, scope),
+        sameOrderExpressions = so.sameOrderExpressions.map(soe => ScopedExpression(soe, scope))
+      )
+    }
+  }
+
+  private def isNotScoped(sortOrder: SortOrder): Boolean =
+    !sortOrder.child.isInstanceOf[ScopedExpression]
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+    // SPARK-42199: sort order of MapGroups must be scoped to their dataAttributes
```

Comment:
We need to revisit the design. We need a way to make sure the attributes inside `ScopedExpression` are resolved by the scoping attributes, but the current design can't guarantee it, as any rule can resolve attributes inside `ScopedExpression` without respecting the scoping attributes.

Reply:
Any suggestion? I lack the knowledge of the codebase to suggest a better design. Who can help here?

Reply:
I think we should look into https://github.com/apache/spark/pull/39754/files#r1747046637 and figure out if there are other rules that need to resolve `ScopedExpression`.